commit 704f3acfcf55343043bbed01c5fb0a0094a68e8a
Author: Denis Pynkin <denis.pynkin@collabora.com>
Date:   Fri Jul 21 19:28:42 2017 +0300

    net: Use packed structures for networking
    
    PXE boot is broken with GCC 7.1 due option '-fstore-merging' enabled
    by default for '-O2':
    
    BOOTP broadcast 1
    data abort
    pc : [<8ff8bb30>]          lr : [<00004f1f>]
    reloc pc : [<17832b30>]    lr : [<878abf1f>]
    sp : 8f558bc0  ip : 00000000     fp : 8ffef5a4
    r10: 8ffed248  r9 : 8f558ee0     r8 : 8ffef594
    r7 : 0000000e  r6 : 8ffed700     r5 : 00000000  r4 : 8ffed74e
    r3 : 00060101  r2 : 8ffed230     r1 : 8ffed706  r0 : 00000ddd
    Flags: nzcv  IRQs off  FIQs off  Mode SVC_32
    Resetting CPU ...
    
    Core reason is usage of structures for network headers without packed
    attribute.
    
    Reviewed-by: Yauheni Kaliuta <yauheni.kaliuta@redhat.com>
    Signed-off-by: Denis Pynkin <denis.pynkin@collabora.com>
    Acked-by: Joe Hershberger <joe.hershberger@ni.com>

Index: karo-tx-uboot/include/net.h
===================================================================
--- karo-tx-uboot.orig/include/net.h
+++ karo-tx-uboot/include/net.h
@@ -274,7 +274,7 @@ struct ethernet_hdr {
 	u8		et_dest[6];	/* Destination node		*/
 	u8		et_src[6];	/* Source node			*/
 	u16		et_protlen;	/* Protocol or length		*/
-};
+} __attribute__((packed));
 
 /* Ethernet header size */
 #define ETHER_HDR_SIZE	(sizeof(struct ethernet_hdr))
@@ -292,7 +292,7 @@ struct e802_hdr {
 	u8		et_snap2;
 	u8		et_snap3;
 	u16		et_prot;	/* 802 protocol			*/
-};
+} __attribute__((packed));
 
 /* 802 + SNAP + ethernet header size */
 #define E802_HDR_SIZE	(sizeof(struct e802_hdr))
@@ -306,7 +306,7 @@ struct vlan_ethernet_hdr {
 	u16		vet_vlan_type;	/* PROT_VLAN			*/
 	u16		vet_tag;	/* TAG of VLAN			*/
 	u16		vet_type;	/* protocol type		*/
-};
+} __attribute__((packed));
 
 /* VLAN Ethernet header size */
 #define VLAN_ETHER_HDR_SIZE	(sizeof(struct vlan_ethernet_hdr))
@@ -333,7 +333,7 @@ struct ip_hdr {
 	u16		ip_sum;		/* checksum			*/
 	struct in_addr	ip_src;		/* Source IP address		*/
 	struct in_addr	ip_dst;		/* Destination IP address	*/
-};
+} __attribute__((packed));
 
 #define IP_OFFS		0x1fff /* ip offset *= 8 */
 #define IP_FLAGS	0xe000 /* first 3 bits */
@@ -361,7 +361,7 @@ struct ip_udp_hdr {
 	u16		udp_dst;	/* UDP destination port		*/
 	u16		udp_len;	/* Length of UDP packet		*/
 	u16		udp_xsum;	/* Checksum			*/
-};
+} __attribute__((packed));
 
 #define IP_UDP_HDR_SIZE		(sizeof(struct ip_udp_hdr))
 #define UDP_HDR_SIZE		(IP_UDP_HDR_SIZE - IP_HDR_SIZE)
@@ -400,7 +400,7 @@ struct arp_hdr {
 	u8		ar_tha[];	/* Target hardware address	*/
 	u8		ar_tpa[];	/* Target protocol address	*/
 #endif /* 0 */
-};
+} __attribute__((packed));
 
 #define ARP_HDR_SIZE	(8+20)		/* Size assuming ethernet	*/
 
@@ -435,7 +435,7 @@ struct icmp_hdr {
 		} frag;
 		u8 data[0];
 	} un;
-};
+} __attribute__((packed));
 
 #define ICMP_HDR_SIZE		(sizeof(struct icmp_hdr))
 #define IP_ICMP_HDR_SIZE	(IP_HDR_SIZE + ICMP_HDR_SIZE)
Index: karo-tx-uboot/net/bootp.h
===================================================================
--- karo-tx-uboot.orig/net/bootp.h
+++ karo-tx-uboot/net/bootp.h
@@ -49,7 +49,7 @@ struct bootp_hdr {
 	char		bp_sname[64];	/* Server host name		*/
 	char		bp_file[128];	/* Boot file name		*/
 	char		bp_vend[OPT_FIELD_SIZE]; /* Vendor information	*/
-};
+} __attribute__((packed));
 
 #define BOOTP_HDR_SIZE	sizeof(struct bootp_hdr)
 
Index: karo-tx-uboot/net/dns.h
===================================================================
--- karo-tx-uboot.orig/net/dns.h
+++ karo-tx-uboot/net/dns.h
@@ -29,7 +29,7 @@ struct header {
 	uint16_t	nauth;		/* Authority PRs */
 	uint16_t	nother;		/* Other PRs */
 	unsigned char	data[1];	/* Data, variable length */
-};
+} __attribute__((packed));
 
 void dns_start(void);		/* Begin DNS */
 
Index: karo-tx-uboot/net/nfs.h
===================================================================
--- karo-tx-uboot.orig/net/nfs.h
+++ karo-tx-uboot/net/nfs.h
@@ -68,7 +68,7 @@ struct rpc_t {
 			uint32_t data[19];
 		} reply;
 	} u;
-};
+} __attribute__((packed));
 void nfs_start(void);	/* Begin NFS */
 
 
Index: karo-tx-uboot/net/sntp.h
===================================================================
--- karo-tx-uboot.orig/net/sntp.h
+++ karo-tx-uboot/net/sntp.h
@@ -51,7 +51,7 @@ struct sntp_pkt_t {
 	unsigned long long originate_timestamp;
 	unsigned long long receive_timestamp;
 	unsigned long long transmit_timestamp;
-};
+} __attribute__((packed));
 
 void sntp_start(void);	/* Begin SNTP */
 
diff --git a/net/Makefile b/net/Makefile
index 6ba8f7575e..a147131918 100644
--- a/net/Makefile
+++ b/net/Makefile
@@ -6,6 +6,7 @@
 #
 
 #ccflags-y += -DDEBUG
+ccflags-y += -Wno-address-of-packed-member
 
 obj-y += checksum.o
 obj-$(CONFIG_CMD_NET)  += arp.o
